### ISSUE TITLE
account anglo saxon - V8

### ISSUE THREAD

**Issue by KevinMcMenamin** [odoo/odoo#2414]
there are multiple issues with anglo-saxon - I have attached a document to the community mail-list, but you can get another copy by emailing me at kevin.mcmenamin@solnet.co.nz

PO for service line 	
Inv control = based on incoming shipments	
Process invoice	
Cannot process as no picking and no option to invoice

PO for service line 	
Inv control = based on PO Lines	
Process invoice	
should be:
DR Expense
CR  Creditors	
Actually:
DR Waiting Supplier Inv
CR   Creditors

PO for both stock & service line 	I
nv control = based on incoming shipments
Service line not on invoice

Return stock 		
Stock Move	
cost should be based on related PO line cost, so should be 
DR Waiting Supplier Inv 5.00
CR  Inventory  5.00	
actually uses average cost, so
DR Waiting Supplier 1.33
CR  Inventory  1.33
Issues:
1. Both the move entry and the refund journal from the credit note is based on the average cost. Should both be based on the original move (since this is what the credit from the supplier will be based on)
2. Invoice posts to ‘to be invoiced’ account not ‘waiting supplier invoice’
3. Same issue with ‘to be invoiced’ irrespective of source PO invoicing method

Direct ship PO for stock line  	
Invoice control =  based on incoming shipments	
Should be	
DR Expense  1.00
CR  Creditors   1.00	
Actually
DR Waiting Supplier Inv
CR   Creditors
Direct ship transactions do create an Incoming Shipments entry, so an alternative to handling via the invoice is to create the moves with the associated journal entries for the out move.
Note: for a direct ship, can set the location to field but cannot specify the customer detail.

Invoice control = based on PO Lines - for direct ship, same result

Creditor invoice without PO – stock product		
Should be	
DR Expense  
CR  Creditors  	
Actually
DR Waiting Supplier Inv
CR   Creditors

Financial Credit	
With stock & service line	Process	
should be
CR Expense(for stock line)
CR Expense
DR Creditors	
actually
CR To be Invoiced (for stock line)
CR Expense
DR Creditors

Sales

Stock Return		
Receive Goods	
should be
DR Stock
CR To Be Invoiced	
actually
DR Stock
CR To Be Invoiced
so stock move is ok
Process Invoice	
should be
DR To Be Invoiced
CR Expense	
actually
DR Waiting Supplier Invoice
CR Expense

Financial Credit		
Process Credit	
should be
No entry for cost	
actually
DR Waiting Supplier Invoice
CR Expense 

Cost Price Adjustment
should be
DR Stock
CR Expense	
actually 
DR Stock
CR To Be Invoiced

**Comment by qdp-odoo** [comment]
i don't know what to say: it seems you made a real ant work by listing all the possible use cases but, honesty, it's not readable at all. :sob: 

i really encourage you to create one issue per wrong use case, and keep the present list in a more readable _table_ depicting for each different use case if it's ok or not (for what version) and the related github issue, if any.

sorry but there's not much i can do for this issue

**Comment by KevinMcMenamin** [comment]
Frankly I find your comment disappointing but not surprising - I have been reporting issues to odoo for 4 years on Anglo-Saxon without getting proper resolution. Since it is clear that you are incapable of resolving the issues yourself I would have thought you would have reached out to someone like myself who could help, but instead you just keep putting your head in the sand and hope it goes away - I can tell you V9 is actually WORSE than V8 !

**Comment by gdgellatly** [comment]
@qdp-odoo - I see what you mean that the issue presented is inactionable.  I couldn't really read it but urge you to engage with Kevin on this.

For years now I have relied on @KevinMcMenamin tests/spreadhseet to validate my own implementations of anglosaxon accounting, even converting the ones that matter to me into automated tests albeit on existing internal data.  It is very complete and correct and comes from a chartered accountant.  For years the community has all created their own different versions because the standard version does not work beyond a very simple, one supported workflow only implementation.

Indeed that supported workflow requires realtime inventory valuation, but now in V9 the recommended is periodical.  I actually cannot even fathom how Anglosaxon will work in this scenario and have started considering a cross between periodical/realtime for anglo-saxon just to make it work where stock is updated at invoice rather than reception or dispatch.

I honestly do not think one issue per wrong case is the correct approach, although perhaps a test suite covering all the cases rather than the list above, I agree.  The reason I do not think one issue per case is right, is that having spent a long time inside this code myself is that fixing one issue at a time patches the symptom rather than the cause.  It needs a holistic overview.

**Comment by qdp-odoo** [comment]
Kevin, the fact I closed _this_ bug report doesn't mean we're refusing to see the problems related to the anglo saxon accounting: given the number of issues related to that topic I fixed since the release of the v9 I'm well placed to know anglo saxon wasn't ready for v9. I just closed this one because it cannot be processed -according to me and to all-the-guys-that-could-have-worked-on-it-since-one-year-but-didn't:  it is not understandable.

If there exist some tests/spreadsheet I'd gladly read that instead. 

@gdgellatly I don't get why you say it's fixing the symptoms rather than the cause... What is the cause, if it's not the different bugs? What needs to be completly changed/reviewed/rewritten that is, according to you, the reason of the different bugs it exists for the anglo saxon accounting?

**Comment by qdp-odoo** [comment]
wait a minute... are you talking about _this_ document, Kevin?

[Anglo-Saxon Test Plan - V8.pdf](https://github.com/odoo/odoo/files/10746/Anglo-Saxon.Test.Plan.-.V8.pdf)

I bet you were! Now your first post starts to make some sense to me. Your use cases, though, are still a bit nebulous. I need to dig more into, and adapt to the v9.

**Comment by KevinMcMenamin** [comment]
pre {display: block;font-family: Helvetica Neue, Helvetica, Arial, sans-serif;white-space: pre-wrap;margin: 1em 0;}It might be useful to you if we worked through with a screen share which I an happy to do

Sent from IBM VerseAttention:
This email may contain information intended for the sole use of
the original recipient. Please respect this when sharing or
disclosing this email's contents with any third party. If you
believe you have received this email in error, please delete it
and notify the sender or postmaster@solnetsolutions.co.nz as
soon as possible. The content of this email does not necessarily
reflect the views of Solnet Solutions Ltd.
qdp-odoo --- Re: [odoo] account anglo saxon - V8 (#2414) --- From:"qdp-odoo" <notifications@github.com>To:"odoo/odoo" <odoo@noreply.github.com>Cc:"Kevin McMenamin" <kevin.mcmenamin@solnet.co.nz>Date:Sat, 10/10/2015 01:34Subject:Re: [odoo] account anglo saxon - V8 (#2414)wait a minute... are you talking about this document, Kevin?

Anglo-Saxon Test Plan - V8.pdf

I bet you were! Now your first post starts to make some sense to me. Your use cases, though, are still a bit nebulous. I need to dig more into, and adapt to the v9.

**Comment by gdgellatly** [comment]
@qdp-odoo I don't know if you remember initial Anglo Saxon submission for
v5 but was basically a  bug fix. Everything since then has built on that
original.  All I mean is that perhaps taking a wider view of why this
module causes so many bugs,  rather than constantly patching it for new
cases might be an alternative approach.

**Comment by gdgellatly** [comment]
As an example there are probably 4 subsets of treatments based on document
origin and stock valuation method.  But this is known at time of creation
and does not change, so it would be far simpler to encode that in a field
on the document,  have a clean dispatch function to allow for future or
unconsidered cases.

**Comment by qdp-odoo** [comment]
@KevinMcMenamin : yes, a real talk will probably come in handy. What is your time shift compared to Brussel's timezone?

So far, I went through the document made for v8 and found 3 recurrent problems:
 1) invoicing based on shipment & service lines (use cases 4,5,7): shouldn't be a problem anymore since there is no invoicing based on shipment. Everything starts from the PO now.

 2) return of products having average as costing method wrongly taking the average price instead of the purchase price for stock valuation entries (use case 10): this is related to this old discussion on Launchpad https://bugs.launchpad.net/openobject-addons/+bug/610738 where no one never found the argument/legal document that makes us beleive we were wrong... though, today I'd tend to agree with you: given that the supplier refund will be based on the purchase price (quite logical) it will credit the stock input account of the purchase price, if we valuate the return shipment at the average price, the stock input account will be wrong because the 2 operations won't cancel each other. We need to 
 * use purchase price for the return
 * recompute the average price after the return

3) vendor bills/refunds not related to a PO should use the expense account instead of the stock input account (use case 13,15): that's the first time I read such a remark. I need to dig that: do you have a reference document for that? Any opinion, people? @gdgellatly ?

Also: 
* use case 10: creating a refund from the original invoice shouldn't change, by default, the account on invoice lines. That should solve the second part of this use case
* use case 11,12: i don't get the use case.
* use case 28: what do you call a financial credit? a sale refund?
* use case 27 and 29: do not seems related to any other issue above. I'll check that.

Thanks

**Comment by KevinMcMenamin** [comment]
time difference is 11hr so early or late your time would work. I'll comment separately on your questions

**Comment by KevinMcMenamin** [comment]
Case 10 - this refers to a stock credit - the two issues are the cost price (should use the originating move cost price and the invoice line is being coded to the output account and should be the input account (ie to offset the journal from the stock move)

**Comment by KevinMcMenamin** [comment]
11 & 12 related to a drop ship from a supplier direct to a customer. I have not tested this properly.

**Comment by KevinMcMenamin** [comment]
28 - a financial credit is one done via accounting rather than stock - so there is no stock move so there should be no entry to an input/output account. Note I have recently fixed a further issue with financial credits - you need to test all three ways in which a financial credit can be generated.

**Comment by KevinMcMenamin** [comment]
Do you want to deal with V9 in this issue or a new one? I created an instance on odoo (solnet.odoo.com) about 1 week ago and here are the test results (won't let me insert a file )

**Comment by gdgellatly** [comment]
I know it seems obvious for the average cost return and I was a big
advocate for it but eventually discovered I was wrong.  Just do some
research on that one first.

**Comment by KevinMcMenamin** [comment]
I don't think that it would be material either way, but the issue to me is what happens in the real world and getting journal entries that balance.
when goods are returned to a supplier the credit that comes would be based on the purchase price of those goods not the current price, so the stock move needs to be at that price or you end up handling a price difference.
when goods are returned by a customer then while I like the idea of having the cost price that relates to that sale line used (as this caters for situation where the cost price was not the average cost say when goods were specifically purchased for that sale) but I can live with average cost.

**Comment by gdgellatly** [comment]
Kevin,

I wish I lived in your real world :).  The thing that led me to search it
out was Fabien's initial request for proof, and at the time I found he was
right.  I wish I could find the reference, it was an IASB doc anyway.  But
trying to find a reference this morning has led me nowhere but into
conflicting information both ways.  In any case, you won't get that past
Fabien.  BTW I am not suggesting that the current situation is right
either.  Below some food for thought.

My real world and why average cost with price difference is appropriate for
supplier returns.

Average cost $2.00
Supplier quotes $1980/ton for some material
PO place for $1980 /kg for 20 tonnes.
New average cost $1979
Invoice comes and mistake seen.
Average Cost corrected and updated in system, new average cost $1.985 - at
this point we have a big counter price difference entered.  CR Inventory DR
Price Difference (or Inventory Loss gain depending on setup) OR as often
happens, a big correction is done to fix up all the mistakes over the year,
same effect.
Some months later - return 1 tonne to supplier.  But we already corrected
the inventory.  Using average cost keeps it all neat and tidy.

In fact if anything the argument should be to adjust average cost by the
original INVOICE amount and apply a price difference.  But then you have
things like restocking fees, landed costs etc.

The other reason is simply this - the average cost method is meant for when
inventory items are not individually identifiable.  So while we may be
nominally returning stock to a supplier, we don't in theory know that it
was exactly that stock.  Sometimes I wish there was a case for - I made a
mistake, just reverse everything no other moves have happened - and then
one for BAU.

**Comment by KevinMcMenamin** [comment]
I did a quick search but the sites don't really deal with the costing issue - just that you adjust for the value of the goods returned.
I would argue on the basis that inventory should be valued on the costs incurred in getting to current condition and location, so whatever the supplier credit is should be the amount that the value of stock is adjusted.
Your situation of an amended average cost does cause an issue as my solution is to use the originating move cost which will not have been updated (which you could do from the cost change wizard).
Case of which option do you want - both end up with some compromise

**Comment by KevinMcMenamin** [comment]
Here is the IAs standard for inventory:

Measurement of inventories
Cost should include all: [IAS 2.10]
costs of purchase (including taxes, transport, and handling) net of trade discounts received costs of conversion (including fixed and variable manufacturing overheads) and other costs incurred in bringing the inventories to their present location and condition

so really what the supplier credits should be the adjustment to inventory cost. Given we don't have this at the time of the move, then the cost that the goods were purchased for has to be the best proxy.

**Comment by gdgellatly** [comment]
Kevin you know me, the option I want is the option.  IOW, in different
countries, companies and settings one, either or both may be equally
valid.  Also to a lesser or greater extent, who am I (or you, or Odoo) to
dictate situation dependent accounting policies.  Anglosaxon as some kind
of global treatment to me fails on the detailed points.  Thats why I
suggested a refactor rather than dealing with only one way all the time and
large monolithic functions to deal with all cases.  So I could plug in,
use_average_cost_for_returns module, someone else could plugin
use_supplier_invoice_for_returns module etc.  Sure have a sensible default
that suits 90% of cases, but make it easy for the 10% to adjust.

Also your treatment, the cost the goods were purchased assumes that the
original purchase for the actual goods are identifiable, in which case IFRS
calls for actual costing anyway.  Which is kind of my point.  I think using
original purchase cost makes sense, until it doesn't.  In fact one could
argue though that this is a missing feature, a plain Return Goods to
Supplier function, irrespective of the original purchases.  That said I
still think the invoice price makes more sense than the move price if only
because the move price can be adjusted afterwards via non-linked cost
updates, whereas the invoice price, financial credits aside really
doesn't.  Perhaps the answer is to just propose a move cost in the return
picking, the credit amount you are expecting.

**Comment by KevinMcMenamin** [comment]
remember when you used to be able to put a cost in on the move wizard.. time to bring it back :)

**Comment by qdp-odoo** [comment]
okay guys, I've taken some time to work this out and I'll try to make a clean summary of what is the behavior of Odoo in which case and why we do that way, as soon as possible.

Kevin, let's wait for that document prior to any screensharing so that we can use that as a basis for future discussions.

**Comment by qdp-odoo** [comment]
hello again anglo saxon fellows,

here is an important update, I'll probably share that in the mailing list as well in the beginning of next week but you guys have the scoop ;-)

1) I think that all problems reported by Kevin in his document are (finally) now ok
2) you can test that on runbot in <b>master</b>. Only some of them are fixed in v9: when you'll read the different things I made, you'll understand that there were quite important behavior changes that couldn't land in a stable version. Good thing is that the next saas release is planned very soon.

The things changed are:
* returned product in average costing method to a supplier now works as described here: https://docs.google.com/document/d/1PsCT28I8zD4los7uivSoWFe0gYbtuWb6tuwja9QL1J4/edit#
 tldr: average price isn't update, the stock journal entry made is using the average price on the stock account, the original purchase price on the stock input account and, if it exists, the difference between goes in the price difference account. I think that's very neat.

(fixes problem n°10)

* the system now only makes stock accounting entries for <b>stockable</b> products. Not anymore for consumable products. Which means than, by side effect, the anglo saxon mode now also only makes a difference for stockable products as well. 
 
That is supposed to fix all the problems related to the stock input account used instead of the expense one when creating a vendor bill without PO (for example). Now it's clear: if you want to keep a stock value for that, use a stockable product and everywhere you'll have the stock input account by default instead of the expense. If you don't want, use a consumable product and the expense account will be used. That means you can also use that for the drop shipping products.

(fixes problems n°11,12,13,15)

* making a refund from scratch follows the same rule, but creating a refund from a vendor bill now keep the accounts on line unchanged.

(fixes problems n°27,28)

* the account used for cost price adjustment has been changed to expense instead of stock input/output

(fixes problems n°29)

* As I said, all the problems related to invoicing based on shipping are gone with the last version of purchase, in v9 (problems n°4,5,7).

That's, more or less, it.

Let me know your thoughts, and if you have time for tests it will be great!


Thanks

**Comment by KevinMcMenamin** [comment]
Just been through my standard test routine - improvements but still some fails plus a couple of things I can't test as I can't work out how to process.
I can't attach my test doc so will email separately.

**Comment by KevinMcMenamin** [comment]
Here is my test results.

Note - I have not done any testing around different currencies which will 
also need to be done once the base issues are resolved.




Kevin McMenamin
ERP Capability Manager

Mobile  +64 22 651 3753
Phone  +64 9 977 5805
Email   kevin.mcmenamin@solnet.co.nz

Solnet Solutions Limited
Level 19, Crombie Lockwood
191 Queen Street, Auckland 1010
PO Box 6619, Auckland 1141

www.solnet.co.nz

**Comment by jbeficent** [comment]
@KevinMcMenamin I'm very interested in the test results. Could you attach them to this github issue? It seems that they were not attached.

**Comment by qdp-odoo** [comment]
[Anglo-Saxon Test Plan - V9 - master 108732.docx](https://github.com/odoo/odoo/files/49504/Anglo-Saxon.Test.Plan.-.V9.-.master.108732.docx)

here's the doc from Kevin he sent me. It's not updated with my last changes, though. 

And, btw, I know there is still a bug on customer returns (they don't zero out the stock output with the customer refund), and we're working on that. Trying to figure out what's the best way of doing that, but it will be done in master anyway because it implies changes in the database schemes.

**Comment by jbeficent** [comment]
@qdp-odoo I realize that some way or another, when there's a refund you need to specify whether this is a credit/debit note (perhaps a flag or selection field in the invoice) (no stock move involved, either to occur or already ocurred), so that the application proposes income/expense accounts instead of stock input/output.

**Comment by jbeficent** [comment]
@qdp-odoo SAP calls those invoices that do not involve a physical movement of goods Credit Memo or Debit Memo. They are classified different from actual invoices because they produce different journal entries, and they are usually related to a prior invoice.

**Comment by passdot** [comment]
Hi Guys, any update on this thread, has all the issues reported been resolved for version 8?

**Comment by TheCloneMaster** [comment]
@qdp-odoo <https://github.com/qdp-odoo> - If you could please give us
pointers to the exact code fixes, so that we can try to adjust it somehow
for V8

Regards,
-Mario

**Comment by passdot** [comment]
@qdp-odoo I am not a developer, as I mentioned, I got it developed from a developer, therefore I will not be able to answer any of your technical queries, however I can share the Module with you if that can help you to update V8.

**Comment by TheCloneMaster** [comment]
All help is welcomed!

**Comment by passdot** [comment]
Please guide me how to upload the module here.
